### PR TITLE
Testable emulator detection

### PR DESCRIPTION
**Pull request by jskeet**
This builds on #4995. Only the second commit is new.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 1]
```diff
@@ -303,15 +303,15 @@ public EmulatorDetection EmulatorDetection
...
+                    (value is EmulatorDetection parsed || (value is string text && Enum.TryParse(text, out parsed))))
                 {
                     return parsed >= EmulatorDetection.None && parsed <= EmulatorDetection.EmulatorOrProduction ? parsed : EmulatorDetection.None;
                 }
                 return EmulatorDetection.None;
             }
             set
             {
-                GaxPreconditions.CheckEnumValue((EmulatorDetection)value, "value");
+                GaxPreconditions.CheckEnumValue(value, "value");
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:314]: nit: nameof(value);

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 2]
```diff
@@ -303,15 +303,15 @@ public EmulatorDetection EmulatorDetection
             get
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:303]: nit: get and set here are not consistent. set fails if the value is out of range but get returns None.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs]: Done.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:303]: That's the same for all the properties. For int ones for example, we have `GetInt32OrDefault`, but `SetInt32WithValidation`.

**Pull request merged**